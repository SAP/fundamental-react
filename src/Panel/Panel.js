import PropTypes from 'prop-types';
import React from 'react';

// ------------------------------------------- Panel ------------------------------------------
export const Panel = props => {
    const { colSpan, children, className, ...rest } = props;
    return <div className={`fd-panel${colSpan ? ' fd-has-grid-column-span-' + colSpan : ''}${className ? ' ' + className : ''}`} {...rest}>{children}</div>;
};

Panel.propTypes = {
    className: PropTypes.string,
    colSpan: PropTypes.number
};

Panel.defaultProps = {
    colSpan: null
};

// ------------------------------------------- Panel Grid ------------------------------------------
export const PanelGrid = props => {
    const { nogap, cols, children, className, ...rest } = props;
    return (
        <div
            className={`fd-panel-grid${nogap ? ' fd-panel-grid--nogap' : ''}${
                cols ? ' fd-panel-grid--' + cols + 'col' : ''
            }${className ? ' ' + className : ''}`} {...rest}>
            {children}
        </div>
    );
};

PanelGrid.propTypes = {
    className: PropTypes.string,
    cols: PropTypes.number,
    nogap: PropTypes.bool
};

PanelGrid.defaultProps = {
    nogap: false,
    cols: null
};

// ------------------------------------------- Panel Body ------------------------------------------
export const PanelBody = props => {
    const { children, className, ...rest } = props;
    return <div className={`fd-panel__body${className ? ' ' + className : ''}`} {...rest}>{children}</div>;
};

PanelBody.propTypes = {
    className: PropTypes.string
};

// ------------------------------------------- Panel Header ------------------------------------------
export const PanelHeader = props => {
    const { children, className, ...rest } = props;
    return <div className={`fd-panel__header${className ? ' ' + className : ''}`} {...rest}>{children}</div>;
};

PanelHeader.propTypes = {
    className: PropTypes.string
};

// ------------------------------------------- Panel Head ------------------------------------------
export const PanelHead = props => {
    const { title, description, className, ...rest } = props;
    return (
        <div className={`fd-panel__head${className ? ' ' + className : ''}`} {...rest}>
            {title ? <h1 className='fd-panel__title'>{title}</h1> : null}
            {description ? <p className='fd-panel__description'>{description}</p> : null}
        </div>
    );
};

PanelHead.propTypes = {
    className: PropTypes.string
};

// ------------------------------------------- Panel Actions ------------------------------------------
export const PanelActions = props => {
    const { children, className, ...rest } = props;
    return <div className={`fd-panel__actions${className ? ' ' + className : ''}`} {...rest}>{children}</div>;
};

PanelActions.propTypes = {
    className: PropTypes.string
};

// ------------------------------------------- Panel Filters ------------------------------------------
export const PanelFilters = props => {
    const { children, className, ...rest } = props;
    return (
        <div className={`fd-panel__filters${className ? ' ' + className : ''}`} {...rest}>
            {children}
        </div>
    );
};

PanelFilters.propTypes = {
    className: PropTypes.string
};


// ------------------------------------------- Panel Content ------------------------------------------
export const PanelContent = props => {
    const { children, className, ...rest } = props;
    return <div className={`fd-panel__content${className ? ' ' + className : ''}`} {...rest}>{children}</div>;
};

PanelContent.propTypes = {
    className: PropTypes.string
};

// ------------------------------------------- Panel Footer ------------------------------------------
export const PanelFooter = props => {
    const { children, className, ...rest } = props;
    return <div className={`fd-panel__footer${className ? ' ' + className : ''}`} {...rest}>{children}</div>;
};

PanelFooter.propTypes = {
    className: PropTypes.string
};
