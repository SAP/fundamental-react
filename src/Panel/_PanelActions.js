import classnames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

const PanelActions = props => {
    const { children, className, ...rest } = props;

    const panelActionsClasses = classnames(
        'fd-panel__actions',
        className
    );

    return <div {...rest} className={panelActionsClasses}>{children}</div>;
};

PanelActions.displayName = 'Panel.Actions';

PanelActions.propTypes = {
    children: PropTypes.node,
    className: PropTypes.string
};

export default PanelActions;
