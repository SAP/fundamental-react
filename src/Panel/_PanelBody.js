import classnames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

const PanelBody = props => {
    const { children, className, ...rest } = props;

    const panelBodyClasses = classnames(
        'fd-panel__body',
        className
    );

    return <div {...rest} className={panelBodyClasses}>{children}</div>;
};

PanelBody.displayName = 'Panel.Body';

PanelBody.propTypes = {
    children: PropTypes.node,
    className: PropTypes.string
};

export default PanelBody;
