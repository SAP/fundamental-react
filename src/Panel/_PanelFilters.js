import classnames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

const PanelFilters = props => {
    const { children, className, ...rest } = props;

    const panelFiltersClasses = classnames(
        'fd-panel__filters',
        className
    );

    return (
        <div {...rest} className={panelFiltersClasses}>
            {children}
        </div>
    );
};

PanelFilters.displayName = 'Panel.Filters';

PanelFilters.propTypes = {
    children: PropTypes.node,
    className: PropTypes.string
};

export default PanelFilters;
