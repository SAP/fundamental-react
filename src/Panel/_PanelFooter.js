import classnames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

const PanelFooter = props => {
    const { children, className, ...rest } = props;

    const panelFooterClasses = classnames(
        'fd-panel__footer',
        className
    );

    return <div {...rest} className={panelFooterClasses}>{children}</div>;
};

PanelFooter.displayName = 'Panel.Footer';

PanelFooter.propTypes = {
    children: PropTypes.node,
    className: PropTypes.string
};

export default PanelFooter;
