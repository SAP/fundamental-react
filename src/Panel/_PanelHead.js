import classnames from 'classnames';
import CustomPropTypes from '../utils/CustomPropTypes/CustomPropTypes';
import PropTypes from 'prop-types';
import React from 'react';

const PanelHead = props => {
    const { title, description, className, headingLevel, ...rest } = props;

    const panelHeadClasses = classnames(
        'fd-panel__head',
        className
    );

    const HeadingTag = `h${headingLevel}`;

    return (
        <div {...rest} className={panelHeadClasses}>
            {title ? <HeadingTag className='fd-panel__title'>{title}</HeadingTag> : null}
            {description ? <p className='fd-panel__description'>{description}</p> : null}
        </div>
    );
};

PanelHead.displayName = 'Panel.Head';

PanelHead.propTypes = {
    className: PropTypes.string,
    description: PropTypes.string,
    headingLevel: CustomPropTypes.range(2, 6),
    title: PropTypes.string
};

PanelHead.defaultProps = {
    headingLevel: 3
};

PanelHead.propDescriptions = {
    description: 'Localized text for the description of the panel.',
    title: 'Localized text for the title of the panel.'
};

export default PanelHead;
