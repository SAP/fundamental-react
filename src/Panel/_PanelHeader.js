import classnames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

const PanelHeader = props => {
    const { children, className, ...rest } = props;

    const panelHeaderClasses = classnames(
        'fd-panel__header',
        className
    );

    return <div {...rest} className={panelHeaderClasses}>{children}</div>;
};

PanelHeader.displayName = 'Panel.Header';

PanelHeader.propTypes = {
    children: PropTypes.node,
    className: PropTypes.string
};

export default PanelHeader;
