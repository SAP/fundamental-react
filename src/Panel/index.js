export { default as Panel } from './Panel.js';
export { default as PanelGrid } from './PanelGrid.js';
