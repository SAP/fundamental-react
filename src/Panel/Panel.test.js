import { Button } from '../Button/Button';
import { mount } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';
import { Panel, PanelActions, PanelBody, PanelContent, PanelFilters, PanelFooter, PanelGrid, PanelHead, PanelHeader} from './Panel';

describe('<Panel />', () => {
    const panel = (
        <Panel>
            <PanelHeader>
                <PanelHead
                    description='Panel Description'
                    title={'Panel Header with Actions'} />
                <PanelActions>
                    <Button compact glyph='add'>
                        Add New Button
                    </Button>
                </PanelActions>
            </PanelHeader>
            <PanelFilters>
                <div>Panel Filters</div>
                <br />
            </PanelFilters>
            <PanelBody>
                <div>Panel Body</div>
            </PanelBody>
            <PanelContent>
                <div>Panel Content</div>
            </PanelContent>
            <PanelFooter>Panel Footer</PanelFooter>
        </Panel>
    );

    const panelWithClass = (
        <Panel className='blue'>
            <PanelHeader className='blue'>
                <PanelHead
                    className='blue'
                    description='Panel Description'
                    title={'Panel Header with Actions'} />
                <PanelActions className='blue'>
                    <Button compact glyph='add'>
                        Add New Button
                    </Button>
                </PanelActions>
            </PanelHeader>
            <PanelFilters className='blue'>
                <div>Panel Filters</div>
                <br />
            </PanelFilters>
            <PanelBody className='blue'>
                <div>Panel Body</div>
            </PanelBody>
            <PanelContent className='blue'>
                <div>Panel Content</div>
            </PanelContent>
            <PanelFooter className='blue'>Panel Footer</PanelFooter>
        </Panel>
    );

    const panelGrid = (
        <PanelGrid className='blue'>
            <Panel colSpan={2}>
                <PanelHead title={'Panel Header with Actions'} />
                <PanelBody>Panel</PanelBody>
            </Panel>
            <Panel>
                <PanelHead description='Panel Description' />
                <PanelBody>Panel</PanelBody>
            </Panel>
            <Panel>
                <PanelHead />
                <PanelBody>Panel</PanelBody>
            </Panel>
            <Panel>
                <PanelBody>Panel</PanelBody>
            </Panel>
            <Panel>
                <PanelBody>Panel</PanelBody>
            </Panel>
        </PanelGrid>
    );

    const panelGridNoGap = (
        <PanelGrid nogap>
            <Panel>
                <PanelBody>Panel</PanelBody>
            </Panel>
            <Panel>
                <PanelBody>Panel</PanelBody>
            </Panel>
            <Panel>
                <PanelBody>Panel</PanelBody>
            </Panel>
            <Panel>
                <PanelBody>Panel</PanelBody>
            </Panel>
            <Panel>
                <PanelBody>Panel</PanelBody>
            </Panel>
        </PanelGrid>
    );

    const panelGridSpan = (
        <PanelGrid cols={2}>
            <Panel>
                <PanelBody>Panel</PanelBody>
            </Panel>
            <Panel>
                <PanelBody>Panel</PanelBody>
            </Panel>
            <Panel>
                <PanelBody>Panel</PanelBody>
            </Panel>
            <Panel>
                <PanelBody>Panel</PanelBody>
            </Panel>
        </PanelGrid>
    );

    test('create panels', () => {
        // create panel
        let component = renderer.create(panel);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        // panel grid
        component = renderer.create(panelGrid);
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        // panel with class
        component = renderer.create(panelWithClass);
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        // panel with span
        component = renderer.create(panelGridSpan);
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        // panel with no gap
        component = renderer.create(panelGridNoGap);
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    describe('Prop spreading', () => {
        test('should allow props to be spread to the Panel component', () => {
            const element = mount(<Panel data-sample='Sample' />);

            expect(
                element.getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });

        test('should allow props to be spread to the PanelGrid component', () => {
            const element = mount(<PanelGrid data-sample='Sample' />);

            expect(
                element.getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });

        test('should allow props to be spread to the PanelBody component', () => {
            const element = mount(<PanelBody data-sample='Sample' />);

            expect(
                element.getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });

        test('should allow props to be spread to the PanelHeader component', () => {
            const element = mount(<PanelHeader data-sample='Sample' />);

            expect(
                element.getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });

        test('should allow props to be spread to the PanelHead component', () => {
            const element = mount(<PanelHead data-sample='Sample' />);

            expect(
                element.getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });

        xtest('should allow props to be spread to the PanelHead component\'s h1 element', () => {
            // TODO: placeholder for this test description once that functionality is built
        });

        xtest('should allow props to be spread to the PanelHead component\'s p element', () => {
            // TODO: placeholder for this test description once that functionality is built
        });

        test('should allow props to be spread to the PanelActions component', () => {
            const element = mount(<PanelActions data-sample='Sample' />);

            expect(
                element.getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });

        test('should allow props to be spread to the PanelFilters component', () => {
            const element = mount(<PanelFilters data-sample='Sample' />);

            expect(
                element.getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });

        test('should allow props to be spread to the PanelFooter component', () => {
            const element = mount(<PanelFooter data-sample='Sample' />);

            expect(
                element.getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });

        test('should allow props to be spread to the PanelContent component', () => {
            const element = mount(<PanelContent data-sample='Sample' />);

            expect(
                element.getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });
    });
});
