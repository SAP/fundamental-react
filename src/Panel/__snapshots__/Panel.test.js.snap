// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Panel /> create panels 1`] = `
<div
  className="fd-panel"
>
  <div
    className="fd-panel__header"
  >
    <div
      className="fd-panel__head"
    >
      <h1
        className="fd-panel__title"
      >
        Panel Header with Actions
      </h1>
      <p
        className="fd-panel__description"
      >
        Panel Description
      </p>
    </div>
    <div
      className="fd-panel__actions"
    >
      <button
        className=" fd-button fd-button--compact sap-icon--add"
        disabled={false}
        selected={false}
      >
        Add New Button
      </button>
    </div>
  </div>
  <div
    className="fd-panel__filters"
  >
    <div>
      Panel Filters
    </div>
    <br />
  </div>
  <div
    className="fd-panel__body"
  >
    <div>
      Panel Body
    </div>
  </div>
  <div
    className="fd-panel__content"
  >
    <div>
      Panel Content
    </div>
  </div>
  <div
    className="fd-panel__footer"
  >
    Panel Footer
  </div>
</div>
`;

exports[`<Panel /> create panels 2`] = `
<div
  className="fd-panel-grid blue"
>
  <div
    className="fd-panel fd-has-grid-column-span-2"
  >
    <div
      className="fd-panel__head"
    >
      <h1
        className="fd-panel__title"
      >
        Panel Header with Actions
      </h1>
    </div>
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__head"
    >
      <p
        className="fd-panel__description"
      >
        Panel Description
      </p>
    </div>
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__head"
    />
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
</div>
`;

exports[`<Panel /> create panels 3`] = `
<div
  className="fd-panel blue"
>
  <div
    className="fd-panel__header blue"
  >
    <div
      className="fd-panel__head blue"
    >
      <h1
        className="fd-panel__title"
      >
        Panel Header with Actions
      </h1>
      <p
        className="fd-panel__description"
      >
        Panel Description
      </p>
    </div>
    <div
      className="fd-panel__actions blue"
    >
      <button
        className=" fd-button fd-button--compact sap-icon--add"
        disabled={false}
        selected={false}
      >
        Add New Button
      </button>
    </div>
  </div>
  <div
    className="fd-panel__filters blue"
  >
    <div>
      Panel Filters
    </div>
    <br />
  </div>
  <div
    className="fd-panel__body blue"
  >
    <div>
      Panel Body
    </div>
  </div>
  <div
    className="fd-panel__content blue"
  >
    <div>
      Panel Content
    </div>
  </div>
  <div
    className="fd-panel__footer blue"
  >
    Panel Footer
  </div>
</div>
`;

exports[`<Panel /> create panels 4`] = `
<div
  className="fd-panel-grid fd-panel-grid--2col"
>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
</div>
`;

exports[`<Panel /> create panels 5`] = `
<div
  className="fd-panel-grid fd-panel-grid--nogap"
>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
</div>
`;
